### ISSUE TITLE
Blank screen after running app for awhile

### ISSUE THREAD

**Issue by freejack** [Telerik-Verified-Plugins/WKWebView#41]
We are using this plugin with famo.us and have been plagued by a blank screen issue since we've started using wkwebview.  I don't believe this is related to the 64-bit issue because the screen blanks out even when we create 64-bit only test builds.

What happens is, after leaving the app open for a while (about an hour on test ipad and iphone we're testing with), the main page will close and only the about:blank page is displayed.

Using the safari profiler, you can see there are two pages.  One is blank and the other is the page of the app.  I'm assuming the blank screen we're seeing is the blank page that remains after the app's page has closed.  I've been unable to determine what is causing the page to close because neither xcode nor safari display any errors when this occurs.

This issue is blocking us from launching our app.  It's difficult to say if this is an issue with the plugin itself of the interaction with famo.us.

**Comment by uipoet** [comment]
@EddyVerbruggen @IjzerenHein Any progress on this issue? Our app will soon be forced to revert back to the old web to avoid the "white screen of death" after running the app in the background for long periods of time.

**Comment by IjzerenHein** [comment]
@uipoet, In our case the white screens were caused by excessive memory usage, causing the separate WebContent process to be terminated and leaving a white screen in the app.
Apologies for not yet submitting a PR with the crash recovery code, I will do this tomorrow. The code still needs a little bit of work through in terms of detecting the crash in all scenarios.

**Comment by uipoet** [comment]
@IjzerenHein Thanks for the additional detail and your willingness to submit a PR. In our case, we haven't found a memory issue. We use a web socket connection for data and may have inadvertently released the latest app without `background fetch` checked. Would be wonderful to at least see the app be able to recover after closing the WKWebView, regardless of the reason.

**Comment by IjzerenHein** [comment]
Hi, I've created a pull request for it.
You can test using this repo: https://github.com/IjzerenHein/WKWebView

To enable, add this preference to config.xml:
```xml
<preference name="RecoverFromCrash" value="true" />
```

**Important:**
The check that is used for checking whether the WkWebView has crashed is very rudimentary and most likely inappropriate for other situations. It checks whether the wkwebview 'title' returns an empty string, and if that is the case, it assumes a crash and starts recovery.

to test:
create a small web-app and create a div with a large image every 200ms or so, and you should see it crash pretty quickly.

**Comment by EddyVerbruggen** [comment]
Hi @IjzerenHein thanks for the PR, much appreciated!

I struggle crashing the webview on my iPhone 6 (device and simulator).

I'm hammering the webview with a new image as quickly as possible with this code (that superlargeimg.jpg is 4000x4000 pixels).

```js
function createImage(src) {
  var div = document.createElement('div');
  var img = document.createElement('img');
  img.src = src;
  div.appendChild(img);
  return img;
}

setInterval(function() {
  document.body.appendChild(createImage("img/superlargeimg.jpg"));
}, 0);
```

Any better crash code someone can share?

**Comment by ekelokorpi** [comment]
Crash recovery is not working properly for me, i can see the "WkWebView crash detected, recovering..." message on log, but it just loads blank page. How i can make it to load index.html?

**Comment by IjzerenHein** [comment]
Hi. I've noticed that when you 'reload' the page from the safari debugger, that the crash recovery kicks in, is that also what you are seeing?

**Comment by ekelokorpi** [comment]
No i got HTML5 game, and for test, it's reloading the page continually. If i let it to run for a while, the WKWebView will crash, and then it just shows blank page.

**Comment by revolunet** [comment]
Hi,
Anyone has any info on the "official" WkWebView memory limits ? 
There used to be a known memory issue with UiWebView but i dont know if its still applicable : http://www.fngtps.com/2010/mobile-safari-image-resource-limit-workaround/

**Comment by robboerman** [comment]
Thanks for the useful knowledge. I thought I was the only one battling the white screen of death. Has anyone figured out a workaround already? It's crazy that Apple keeps the WKWebView in this useless state.

**Comment by revolunet** [comment]
We discovered that WkWebView has a low GPU limit. If you upload more than 4096x4096px to the GPU, your app will crash without notice. So use `translate3d` sparingly.

**Comment by robboerman** [comment]
Thanks Julien, I will watch out for that. The other white screens are not related to GPU issues I guess. the WKWebView.WebContent just gets stopped when iOS issues a memory warning.

**Comment by michaelcueno** [comment]
Not sure if you guys were aware, but there is an undocumented (but public) callback on WKNavigationDelegate in iOS 9 that is invoked when the web process is terminated: 

- (void)webViewWebContentProcessDidTerminate:(WKWebView *)webView

This might be a better solution than the WKWebView.title approach.

**Comment by EddyVerbruggen** [comment]
@michaelcueno That looks interesting indeed. I don't know if others have considered it. I didn't because I have never been able to (re)produce a crash, unfortunately.

**Comment by sikko** [comment]
Same issue here... The app goes to blank screen while or just after using the camera (using a <input type=file capture=camera>). We really had big hopes 
Here is the memory usage of the app. The spike is where I take a picture.
![screen shot 2016-03-31 at 13 34 40](https://cloud.githubusercontent.com/assets/951412/14175743/636d62be-f74d-11e5-93e5-1293647fff64.png)

Any solutions yet?

**Comment by robboerman** [comment]
We have had the same issue and have not been able to resolve it. The problem is that the WKWebview process runs on a separate thread from the main thread. That's also why your main proces appears to use so little memory, the DOM stuff is all on the WebView thread. When the app runs into memory problems it will kill background threads, including the webview thread. So whey you return from the camera thread to the webview, its' gone. I have no idea how to solve that. If anyone does, let me know

**Comment by EddyVerbruggen** [comment]
Same here. What may help is having the camera take a smaller / more compressed image but I haven't been able to reproduce this crash with any setting really. Does your app reproduce the crash reliably?

Have you enable crash recovery to avoid the white screen?

**Comment by sikko** [comment]
No, it's not consistent. Most of the time it crashes, but sometimes not... I've also noticed that it is less likely to crash with lighter apps. I've tried this url: https://mobilehtml5.org/ts/?id=23 couldn't get it to crash.  My app is a quite heavy angular app. I tried optimising it but still not enough I guess.

I can't really rely on a crash recovery mechanism. The process before reaching the "take a photo" step is long and tedious. Recovering would mean a complete reload and loss of unsaved data wouldn't it?

**Comment by EddyVerbruggen** [comment]
Yes, unless you'd save the current state in LS before going to the camera. But I'm not advocating that of course. Ironically al I want to do is make it crash consistently so we can figure out a fix.

**Comment by sikko** [comment]
Ok, I don't know if this will help, but I get pretty good crash rate with this app I just forked, and maybe you could use it:

https://jsfiddle.net/sikko/q1nw90Lw/5/show/

It's an angular page with 10,000 watchers and a file input to trigger the camera. Most of the time, after taking the picture, it crashes (on my iPad mini at least).

**Comment by gregavola** [comment]
@EddyVerbruggen Any update on this issue?

**Comment by anymos** [comment]
Any update on this issue ?

**Comment by liuminqian** [comment]
@anymos

I create WebContent Process crash scene like this on iPhone 6 Plus:  a website use many high resolution images and input label with camera,  first open demo and click input label, then also open safari which open the website and click input label, back to the demo finally.

Solution: 
1. iOS >= 9.0, WKNavigationDelegate webViewWebContentProcessDidTerminate to detect WebContent Process Crash, and eval document.title too. To avoid crash recovery again and again, reload webview in viewWillAppear when crash was detected.
2. iOS < 9.0, just eval document.title in viewWillAppear, then reload WKWebView.

**Comment by jpike88** [comment]
I write apps that heavily rely on a form 'engine', that makes use of photos.

This issue is 100% related to RAM usage and how iOS deals with WKWebView (which I think is idiotic, as you can't even intercept a memory warning because it runs as a seperate process!)

The **only** way I'm been able to somewhat offset this issue is why running a semi-frequent autosave mechanism, and also ask users to ensure that they save often when dealing with huge forms.

Worst case scenario is that users get a reload and have their data almost 100% intact.

There is no way around this, it's ultimately the price you pay when building intensive hybrid apps for iOS. The operating system is very anal and kills the WKWebView process with extreme prejudice, so when the view reloads (as per @liuminqian's suggestions), just ensure that the app redirects and reloads their data to make the experience as smooth as possible.

Only 'solutions' to this problem are:

- Encourage users to either use android devices, or latest gen iPhone/iPad combinations. Only devices with more than 1GB of RAM will run the risk of this issue occurring, anything beyond that I find to be alright.
- Perhaps force a 'reload' somehow on a daily basis (I haven't experimented with this as it hasn't been a major issue
- Ensure your UI is as light as possible. Things like heavy CSS, poorly written JS, or excessive use of large images will freak out the memory usage.

**Comment by mikkelsjolin** [comment]
In my meteor app the only "workaround" I have found is making sure "CordovaWebViewEngine" is set to "CDVUIWebViewEngine". But that greatly increases memory and cpu usage.

**Comment by guilllaumeCritizr** [comment]
> Not sure if you guys were aware, but there is an undocumented (but public) callback on WKNavigationDelegate in iOS 9 that is invoked when the web process is terminated:
> 
> `- (void)webViewWebContentProcessDidTerminate:(WKWebView *)webView`
> 
> This might be a better solution than the WKWebView.title approach.

Still a valid point to this date ;) 
Thank you.

**Comment by XandiKarlen** [comment]
> Same issue here... The app goes to blank screen while or just after using the camera (using a input type=file capture=camera). We really had big hopes in wkwebview, but this issue may make us go towards a real native app..
> Here is the memory usage of the app. The spike is where I take a picture.
> ![screen shot 2016-03-31 at 13 34 40](https://cloud.githubusercontent.com/assets/951412/14175743/636d62be-f74d-11e5-93e5-1293647fff64.png)
> 
> Any solutions yet?

Did anyone find a solutions to this by any chance? Because we're currently experiencing the exact same issue with our app...

**Comment by jpike88** [comment]
@XandiKarlen there is so solution, just mitigation. Can you give more details as to your specific use case?

**Comment by jpike88** [comment]
try this instead

https://github.com/apache/cordova-plugin-wkwebview-engine

**Comment by XandiKarlen** [comment]
> try this instead
> 
> https://github.com/apache/cordova-plugin-wkwebview-engine

@jpike88 That's the one we're currently using, but unfortunately we're still facing this weird white-screen issue. We first encountered this when using a partners third party library for scanning documents but were able to reproduce this even with https://cordova.apache.org/docs/en/latest/reference/cordova-plugin-camera/. We spent a lot of time debugging this with no real new insight than what has already been listed here.

**Comment by XandiKarlen** [comment]
At the moment we're playing around with restarting our app and somehow restore our previously used session, but that turns out to be harder than we initially thought 😅  So I was wondering whether others had already a working solution for this in place

**Comment by jpike88** [comment]
I would review whether or not this platform is right for your use case.

In our case we have very complex forms which makes angular a great solution. We have autosave, and are piping image data in base64 across the native bridge a lot as autosave ticks over, only in some edge use cases where over 50+ images at 300x400 resolution are being included in a particular session. We have the mechanism that jumps the user right back into the form in the case of unexpected termination.

This is adequate for us, if your situation deviates from this it’s worth thinking about why and if this is a suitable way to go

**Comment by XandiKarlen** [comment]
You're right about that. In retrospective, we probably wouldn't use a cordova based framework for the type of application we know have.

All in all our use case is similar to yours. We're using a polymer based app utilising a native plugin to scan documents and returning the filepath to the app afterwards. Since upgrading to WKWebView our app now oftentimes runs into this white-screen after taking the picture.

For now we will keep on trying to somehow restore the previous session as a workaround and see where this will get us. @jpike88 Thanks for your quick response and insight to your use case! 👍
